### ISSUE TITLE
Guard is not effective on inlined function 

### ISSUE THREAD

**Issue by c-rack** [elixir-lang/elixir#3249]
The function `Integer.to_char_list/2` is inlined by compiler and has the guard `when is_integer(base) and base in 2..36`:
https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/integer.ex#L244-L247

Now, we would expect that calling `Integer.to_char_list/2` with an invalid argument for `base` should raise a `FunctionClauseError`, because there is no other matching function.

Instead, it seems the guard is ignored and the invalid argument is used, so that the following test case fails (raises `ArgumentError` instead of `FunctionClauseError`).

```elixir
  test "Invalid parameter, inlined function" do
	assert_raise FunctionClauseError, fn ->
		Integer.to_char_list(10, :invalid)
	end
  end
```

A full example project can be found here:
https://github.com/c-rack/elixir-guard-test

I have tested with current Elixir master branch as of 6th of April 2015, last commit 5807274f44c592e580daca5305e084bbb8ee1d72.

**Comment by josevalim** [comment]
Thank you! Closed in favor of #3248.

**Comment by c-rack** [comment]
Sure, when I reported this problem on IRC about 3 hours ago, I was not aware that the community would pick up this problem that fast :+1:
